### PR TITLE
General changes [INDEV]

### PR DESCRIPTION
**Pull request by OfficialNoob**
**General changes :D**
- Added support for channel trailers
- URL is optional when in contentscript
- Global videoID in contentscript
- "v" is now "video"
- added createButton function
- new locales for player buttons
- Using more arrow functions
- Fixed popup.html being hard coded 
- No longer runs getYouTubeVideoID every time
- Added a Wait function that resolves a promise on a function also includes a timeout

### CODE REVIEW

**Comment by OfficialNoob** [comment]
@ajayyy ready for review

File: utils.js [hunk 1]
```diff
@@ -1,23 +1,37 @@
...
+        url = document.URL;
+    }
     //Attempt to parse url
-    let urlObject = null;
-    try { 
-        urlObject = new URL(url);
-    } catch (e) {      
-        console.error("[SB] Unable to parse URL: " + url);
-        return false
+    if(typeof url === 'string') {
```
Comment (ajayyy) [inline utils.js]: What happens if this is false?

File: utils.js [hunk 2]
```diff
@@ -1,23 +1,37 @@
...
+            url = new URL(url);
+        } catch (e) {      
+            console.error("[SB] Unable to parse URL: " + url);
+            return false
+        }
     }
   
     //Check if valid hostname
-    if(!["www.youtube.com","www.youtube-nocookie.com"].includes(urlObject.host)) return false; 
+    if(!["www.youtube.com","www.youtube-nocookie.com"].includes(url.host)) return false;
```
Comment (ajayyy) [inline utils.js]: Why did you change it from the two variables of different types having different names?

File: utils.js [hunk 3]
```diff
@@ -1,23 +1,37 @@
+function getTrailerID() { // Requires DOM level access
+    let p = document.location.pathname;
+    if(!document.location.origin === "https://www.youtube.com") return false
+    if(!p.startsWith("/user/")  && !p.startsWith("/channel/"))  return false
+    if(document.getElementsByClassName("ytp-title-link")[1] === undefined) return false
+    let id = document.getElementsByClassName("ytp-title-link")[1].href.split("?v=")[1];
```
Comment (ajayyy) [inline utils.js]: Interesting method of capture

File: utils.js (hunk 1 above)
Comment (OfficialNoob) [reply utils.js]: the parsing will be skipped this was done to allow url objects to be parsed to it directly but never got used

File: utils.js (hunk 2 above)
Comment (OfficialNoob) [reply utils.js]: "this was done to allow url objects to be parsed to it directly but never got used"

File: utils.js (hunk 3 above)
Comment (OfficialNoob) [reply utils.js]: yeah

File: utils.js (hunk 2 above)
Comment (OfficialNoob) [reply utils.js]: Changed

File: utils.js (hunk 1 above)
Comment (OfficialNoob) [reply utils.js]: Removed check

**Comment by OfficialNoob** [comment]
@ajayyy is it ready?

**Comment by ajayyy** [comment]
Looks good, can you just do some testing now?

**Comment by ajayyy** [comment]
Why is `getYouTubeVideoURL` in `utils.js` if it can run `document.URL`? Have all references to it in other JS files been removed?

**Comment by OfficialNoob** [comment]
@ajayyy only the contentscript references have been removed url support should still work

**Comment by OfficialNoob** [comment]
@ajayyy please fix it :/

**Comment by ajayyy** [comment]
fix what

File: content.js [hunk 4]
```diff
@@ -95,18 +93,24 @@ chrome.storage.sync.get(["dontShowNoticeAgain"], function(result) {
...
-  
+
+function initVideo() {
+	let id = getYouTubeVideoID();
+	if(id !== false) return videoIDChange(id);
+	Wait(getYouTubeVideoID_ALT).then((result) => {
+		if(result !== false) {
+			videoIDChange(result);
+		} else {
+			resetValues();
```
Comment (ajayyy) [inline content.js]: Why not reset values before waiting?

File: content.js [hunk 5]
```diff
@@ -134,7 +138,7 @@ function messageListener(request, sender, sendResponse) {
 
     if (request.message == "getVideoID") {
       sendResponse({
-        videoID: getYouTubeVideoID(document.URL)
+        videoID: videoID
```
Comment (ajayyy) [inline content.js]: Can you init videoID at the top. `var videoID = false`

File: content.js [hunk 6]
```diff
@@ -219,7 +219,7 @@ function resetValues() {
   //reset sponsor times
   sponsorTimes = null;
   UUIDs = null;
-  sponsorVideoID = id;
+  sponsorVideoID = (videoID === undefined) ? false : videoID;
```
Comment (ajayyy) [inline content.js]: Change this line once videoID is set to false by default.

File: content.js [hunk 7]
```diff
@@ -230,8 +230,10 @@ function resetValues() {
 }
 
 function videoIDChange(id) {
+  if(id === false) return false;
+  videoID = id; // Global
   //not a url change
-  if (sponsorVideoID == id) return;
+  if (sponsorVideoID == videoID) return;
```
Comment (ajayyy) [inline content.js]: `sponsorVideoID` and `videoID`? A little confusing having both.

File: utils.js [hunk 8]
```diff
@@ -1,4 +1,36 @@
-function getYouTubeVideoID(url) {
+// Function that can be used to wait for a condition before returning
+async function Wait(condition, timeout = 5000, check = 100) {
```
Comment (ajayyy) [inline utils.js]: Can you change this to camelCase. `wait()`

Comment (OfficialNoob) [reply utils.js]: Its one word is camelCase needed

Comment (ajayyy) [reply utils.js]: yes

File: content.js (hunk 4 above)
Comment (OfficialNoob) [reply content.js]: to give getYouTubeVideoID_ALT a chance

File: content.js (hunk 5 above)
Comment (OfficialNoob) [reply content.js]: sure

File: utils.js (hunk 8 above)
Comment (OfficialNoob) [reply utils.js]: Changed

File: content.js (hunk 7 above)
Comment (OfficialNoob) [reply content.js]: There both called videoID now the id check is no longer using videoID

File: content.js (hunk 6 above)
Comment (OfficialNoob) [reply content.js]: Done

**Comment by ajayyy** [comment]
I'll leave a more detailed review later, but for now I'd like to enforce the current style. Arrow functions should only be used for functions that are one line (without curly brackets)

**Comment by OfficialNoob** [comment]
To many changes may be easier just to start over

File: content.js [hunk 9]
```diff
@@ -255,93 +255,94 @@ function videoIDChange(id) {
...
-
-    updateVisibilityOfPlayerControlsButton();
-  });
-  chrome.storage.sync.get(["hideDeleteButtonPlayerControls"], function(result) {
-    if (result.hideDeleteButtonPlayerControls != undefined) {
-      hideDeleteButtonPlayerControls = result.hideDeleteButtonPlayerControls;
-    }
+    videoID: videoID
+}, (response) => {
+    wait(getVideo).then(result => {
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: The code did not work anyway. I was trying to find the problem

[the remark this answers is not in the document]
Comment (ajayyy) [reply content.js]: Oh okay, makes sense

File: content.js [hunk 10]
```diff
@@ -620,157 +611,46 @@ function updateSponsorTimesSubmitting() {
...
 
   if (showStartSponsor) {
     showingStartSponsor = true;
     document.getElementById("startSponsorImage").src = chrome.extension.getURL("icons/PlayerStartIconSponsorBlocker256px.png");
-    document.getElementById("startSponsorButton").setAttribute("title", chrome.i18n.getMessage("sponsorStart"));
+    document.getElementById("startSponsor").setAttribute("title", chrome.i18n.getMessage("sponsorStart"));
 
     if (document.getElementById("startSponsorImage").style.display != "none" && uploadButtonVisible && !hideInfoButtonPlayerControls) {
-      document.getElementById("submitButton").style.display = "unset";
+      document.getElementById("submitButton").style.visibility = "visible";
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: Well you dont want style.visibility = "hidden"

[the remark this answers is not in the document]
Comment (ajayyy) [reply content.js]: Sorry, I mistyped

File: content.js [hunk 11]
```diff
@@ -520,95 +523,83 @@ function skipToTime(v, index, sponsorTimes, openNotice) {
...
-  }
-  if (hideDeleteButtonPlayerControls) {
-    document.getElementById("deleteButton").style.display = "none";
-  }
+getControls = () => {
+	  let controls = document.getElementsByClassName("ytp-right-controls");
+	  return (controls) ? controls[controls.length - 1] : false;	  
+};
+
+createButton = (baseid, title, callback, imageName) => {
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: The code was not working anyway.

File: background.js [hunk 12]
```diff
@@ -1,7 +1,9 @@
 chrome.tabs.onUpdated.addListener(function(tabId, changeInfo, tab) {
-	chrome.tabs.sendMessage(tabId, {
-		message: 'update',
-	});
+	if(changeInfo.status == "complete") {
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply background.js]: I did not get to testing it

File: content.js [hunk 13]
```diff
@@ -620,157 +611,46 @@ function updateSponsorTimesSubmitting() {
...
-  if (document.getElementById("sponsorBlockPopupContainer") != null) {
-    //it's already added
-    return;
-  }
-
   popupInitialised = false;
 
   //hide info button
-  document.getElementById("infoButton").style.display = "none";
+  document.getElementById("infoButton").style.visibility = "hidden";
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: If the user does not want buttons the html can just not be added the button should not be removed normally display = "none"; would mean the buttons could move

File: content.js [hunk 14]
```diff
@@ -620,157 +611,46 @@ function updateSponsorTimesSubmitting() {
...
 
-  sendRequestToCustomServer('GET', chrome.extension.getURL("popup.html"), function(xmlhttp) {
+  sendRequestToCustomServer('GET', chrome.extension.getURL(browser.i18n.getMessage("Popup")), (xmlhttp) => {
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: :(

File: content.js [hunk 15]
```diff
@@ -1207,7 +1069,7 @@ function getSponsorTimesMessage(sponsorTimes) {
 }
 
 //converts time in seconds to minutes:seconds
-function getFormattedTime(seconds) {
+getFormattedTime = (seconds) => {
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: Closed

File: content.js [hunk 16]
```diff
@@ -211,28 +209,30 @@ document.onkeydown = function(e){
   }
 }
 
-function resetValues() {
+resetValues = () => {
+  // Reset videoID
+  videoID = false;
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: Never understand what you meant :(

File: content.js [hunk 17]
```diff
@@ -835,24 +713,21 @@ function clearSponsorTimes() {
...
 }
 
 //Opens the notice that tells the user that a sponsor was just skipped
-function openSkipNotice(UUID){
-  if (dontShowNotice) {
-    //don't show, return
-    return;
-  }
+openSkipNotice = (UUID) => {
+  if (dontShowNotice) return;// don't show, return
```
[the remark this answers is not in the document]
Comment (OfficialNoob) [reply content.js]: ok

File: content.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (ajayyy) [reply content.js]: That's just a small issue, it just has to be added to the messages.json

File: content.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (ajayyy) [reply content.js]: videoIDChanged checks to see if `videoID` == the new video id. If they are the same, it doesn't run the code. You can resetting `videoID` before calling `videoIDChanged`, so this check will not get triggered.
